### PR TITLE
Make expand_to_rows, expand_column support Rows, Tables, Column data types

### PR DESCRIPTION
**Pull request by AdRiley**
### Pull Request Description

Make expand_to_rows work for Table and Column
Make expand_column work for Column and Row

Makes solving the book club exercise easier

image

### Important Notes

We decided expand_column did not make sense for Table as the resulting Table of Columns would rarely be what was wanted.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Review by jdunkerley — APPROVED** [review]

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Conversions/Convertible_To_Columns.enso [hunk 1]
```diff
@@ -61,3 +72,8 @@ Convertible_To_Columns.from (that:XML_Element) =
     value = if that_children.is_empty.not && has_child_nodes.not then [["Value", that.text]] else []
     as_map = Map.from_vector (name + tags + children + value)
     Convertible_To_Columns.from as_map
+
+## PRIVATE
+index_get that index:Text =
+    val = Integer.parse index
+    if val.is_error then Nothing else that.get val
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Conversions/Convertible_To_Columns.enso]: Why is this `is_error` path needed?

I assumed that given that `fields` are generated from a `Range` mapped with `.to_text`, `Integer.parse` should _always_ succeed in there.

When can it fail?

Comment (AdRiley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Data/Conversions/Convertible_To_Columns.enso]: ~~I think when the things being expanded are ragged...~~

~~See the Nothing in the "Expands columns within a column" test.~~

Nope I am wrong. @jdunkerley We added this together. If I remove it then no tests fail.

Comment (AdRiley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Data/Conversions/Convertible_To_Columns.enso]: Removed it

File: test/Table_Tests/src/In_Memory/Table_Conversion_Spec.enso [hunk 2]
```diff
@@ -206,6 +206,20 @@ add_specs suite_builder =
             table = Table.new [["aaa", [1, 2]], ["bbb", [Date.new 2020 12 1 . up_to (Date.new 2020 12 3), Date.new 2022 12 1 . up_to (Date.new 2022 12 2)]]]
             expected = Table.new [["aaa", [1, 2]], ["bbb 0", [Date.new 2020 12 1, Date.new 2022 12 1]], ["bbb 1", [Date.new 2020 12 2, Nothing]]]
             table.expand_column "bbb" . should_equal expected
+        
+        group_builder.specify "Expands rows within a column" <|
+            table1 = Table.new [["aaa", [1, 2, 3]], ["bbb", [4, 5, 6]], ["ccc", [7, 8, 9]]]
+            table2 = Table.new [["aaa", [10]], ["bbb", [11]], ["ccc", [12]]]
+            table = Table.new [["tables", [table1, table2]]]
+            expected = Table.new [["tables aaa", [1, 2, 3, 10]], ["tables bbb", [4, 5, 6, 11]], ["tables ccc", [7, 8, 9, 12]]]
+            table.expand_to_rows "tables" . expand_column "tables" . should_equal expected
```
Comment (radeusgd) [inline test/Table_Tests/src/In_Memory/Table_Conversion_Spec.enso:215]: shall we have an example where the tables do not have the same sets of columns?

**Review by radeusgd — APPROVED** [review]
Looks good.

It's good to see that the `Convertible_To_Rows` and its sibling allow for such easy extensibility of this feature. It's nice to see some cool idioms for abstraction in Enso settling in.

I think we may want to add one test, although I assume it may not be strictly necessary because it is tested by other paths (I think JSON already tests this codepath well).

File: test/Table_Tests/src/In_Memory/Table_Conversion_Spec.enso (hunk 2 above)
Comment (AdRiley) [reply test/Table_Tests/src/In_Memory/Table_Conversion_Spec.enso:215]: Done

**Review by GregoryTravis — APPROVED** [review]

**Pull request merged**